### PR TITLE
Implement methods to transactionally get and set multiple items

### PR DESCRIPTION
**Pull request by marcinwasowicz**
This PR modifies localforage to support transactional multiple item retrieval/persistence. The synchronizationKey and synchronizationValue arguments are used to handle race conditions. The setMultipleItems call will only succeed if current value under synchronizationKey in localforage is equal to expectedSynchronizationValue or forceWrite is set to true. This lets us select on thread to always win race condition.

This PR implements a strange pattern - creating new IndexedDB request from the success callback of another request instead of creating them in parallel in a loop. Apparently this is the correct way to keep the transaction alive until all request complete: https://stackoverflow.com/questions/10385364/how-do-you-keep-an-indexeddb-transaction-alive.

Two methods mentioned above are implemented only for IndexedDB driver.

### CODE REVIEW

File: typings/localforage.d.ts [hunk 1]
```diff
@@ -17,8 +17,12 @@ interface LocalForageOptions extends LocalForageDbInstanceOptions {
 interface LocalForageDbMethodsCore {
     getItem<T>(key: string, callback?: (err: any, value: T | null) => void): Promise<T | null>;
 
+    getMultipleItems<T>(keys: ReadonlyArray<keyof T>, synchronizationKey: string): Promise<Readonly<{...T, synchronizationValue?: string}>>;
+
     setItem<T>(key: string, value: T, callback?: (err: any, value: T) => void): Promise<T>;
 
+    setMultipleItems<T>(input: T, synchronizationKey: string, expectedSynchronizationValue?: string, newSynchronizationValue: string, forceWrite: boolean): Promise<void>;
```
Comment (xsanm) [inline typings/localforage.d.ts]: I think the `T` here can be typed better, as we always expect this to be an object where keys are IndexedDB keys

File: typings/localforage.d.ts [hunk 2]
```diff
@@ -17,8 +17,12 @@ interface LocalForageOptions extends LocalForageDbInstanceOptions {
 interface LocalForageDbMethodsCore {
     getItem<T>(key: string, callback?: (err: any, value: T | null) => void): Promise<T | null>;
 
+    getMultipleItems<T>(keys: ReadonlyArray<keyof T>, synchronizationKey: string): Promise<Readonly<{...T, synchronizationValue?: string}>>;
```
Comment (xsanm) [inline typings/localforage.d.ts]:
```suggestion
    getMultipleItems<T>(keys: ReadonlyArray<keyof T>, synchronizationKey: string): Promise<Readonly<{value: T, synchronizationValue?: string}>>;
```

same as [here](https://phab.comm.dev/D12836#inline-74725) - we think that it's better and more readable than spreading

File: src/drivers/indexeddb.js [hunk 3]
```diff
@@ -746,6 +746,199 @@ function removeItem(key, callback) {
...
+                        return reject(err);
+                    }
+
+                    try {
+                        var store = transaction.objectStore(
+                            self._dbInfo.storeName
+                        );
+
+                        var allKeys = [...normalizedKeys, synchronizationKey];
+                        var requests = [];
```
Comment (xsanm) [inline src/drivers/indexeddb.js]: I don't think we need this at all

File: src/drivers/indexeddb.js [hunk 4]
```diff
@@ -746,6 +746,199 @@ function removeItem(key, callback) {
...
+                    if (err) {
+                        return reject(err);
+                    }
+
+                    try {
+                        var store = transaction.objectStore(
+                            self._dbInfo.storeName
+                        );
+
+                        var allKeys = [...normalizedKeys, synchronizationKey];
```
Comment (xsanm) [inline src/drivers/indexeddb.js]:
```suggestion
                        var allKeys = [...normalizedKeys, normalizedSynchronizationKey];
```

File: src/drivers/indexeddb.js [hunk 5]
```diff
@@ -746,6 +746,199 @@ function removeItem(key, callback) {
...
+                                var value = req.result;
+                                if (value === undefined) {
+                                    value = null;
+                                }
+
+                                if (_isEncodedBlob(value)) {
+                                    value = _decodeBlob(value);
+                                }
+
+                                if (key === synchronizationKey) {
```
Comment (xsanm) [inline src/drivers/indexeddb.js]:
```suggestion
                                if (key === normalizedSynchronizationKey) {
```

File: src/drivers/indexeddb.js [hunk 6]
```diff
@@ -746,6 +746,199 @@ function removeItem(key, callback) {
...
+
+                                result['values'] = values;
+                                resolve(result);
+                            };
+
+                            req.onerror = function() {
+                                reject(req.err);
+                            };
+
+                            requests.push(req);
```
Comment (xsanm) [inline src/drivers/indexeddb.js]: what is this doing?

File: src/drivers/indexeddb.js [hunk 7]
```diff
@@ -746,6 +746,199 @@ function removeItem(key, callback) {
...
+                                } else {
+                                    values[key] = value;
+                                }
+
+                                if (index < keysArray.length - 1) {
+                                    placeRequest(keysArray, index + 1);
+                                    return;
+                                }
+
+                                result['values'] = values;
```
Comment (xsanm) [inline src/drivers/indexeddb.js:804]: this is different than the TS type - could you fix it? Also see suggestions about the type

File: src/drivers/indexeddb.js [hunk 8]
```diff
@@ -746,6 +746,199 @@ function removeItem(key, callback) {
...
+                    inputArray.map(function([key, value]) {
+                        var normalizedKey = normalizeKey(key);
+                        if (value === null) {
+                            value = undefined;
+                        }
+
+                        if (!toString.call(value) === '[object Blob]') {
+                            return new Promise(function(resolve, reject) {
+                                resolve([normalizedKey, value]);
+                            });
```
Comment (xsanm) [inline src/drivers/indexeddb.js]: You can just return `[normalizedKey, value]` without wrapping in `new Promise` and `Promise.all`  should handle this

File: src/drivers/indexeddb.js [hunk 9]
```diff
@@ -746,6 +746,199 @@ function removeItem(key, callback) {
...
+                            blobSupport
+                        ) {
+                            if (blobSupport) {
+                                return [normalizedKey, value];
+                            }
+                            return [normalizedKey, _encodeBlob(value)];
+                        });
+                    })
+                );
+                return inputNormalizationPromise;
```
Comment (xsanm) [inline src/drivers/indexeddb.js]:
```suggestion
                return Promise.all(
                    inputArray.map(function([key, value]) {
                        var normalizedKey = normalizeKey(key);
                        if (value === null) {
                            value = undefined;
                        }

                        if (!toString.call(value) === '[object Blob]') {
                            return new Promise(function(resolve, reject) {
                                resolve([normalizedKey, value]);
                            });
                        }

                        return _checkBlobSupport(dbInfo.db).then(function(
                            blobSupport
                        ) {
                            if (blobSupport) {
                                return [normalizedKey, value];
                            }
                            return [normalizedKey, _encodeBlob(value)];
                        });
                    })
                );
```

File: src/drivers/indexeddb.js [hunk 10]
```diff
@@ -746,6 +746,199 @@ function removeItem(key, callback) {
...
+                    if (err) {
+                        reject(err);
+                    }
+
+                    try {
+                        var store = transaction.objectStore(
+                            self._dbInfo.storeName
+                        );
+
+                        var writeRequests = [];
```
Comment (xsanm) [inline src/drivers/indexeddb.js]: same as above - I don't think we need this

File: src/drivers/indexeddb.js [hunk 11]
```diff
@@ -746,6 +746,199 @@ function removeItem(key, callback) {
...
+                            var value = req.result;
+                            if (value === undefined) {
+                                value = null;
+                            }
+                            if (
+                                value !== expectedSynchronizationValue &&
+                                !forceWrite
+                            ) {
+                                reject('Another thread completed transaction');
+                            }
```
Comment (xsanm) [inline src/drivers/indexeddb.js]: you not throwing - this is just calling `reject` callback, to make it work you should return here

**Review by xsanm — CHANGES_REQUESTED** [review]
Overall idea looks okay, requesting to address comments but those are mostly related to code quality, not to approach

File: typings/localforage.d.ts (hunk 1 above)
Comment (marcinwasowicz) [reply typings/localforage.d.ts]: The `T` can be object with any keys so I am not sure how to make it more narrow. What do you mean by "object where keys are IndexedDB keys"?

File: src/drivers/indexeddb.js (hunk 6 above)
Comment (marcinwasowicz) [reply src/drivers/indexeddb.js]: As you suggested in one of previous comments it is probably not necessary and can be safely deleted.

File: src/drivers/indexeddb.js (hunk 10 above)
Comment (marcinwasowicz) [reply src/drivers/indexeddb.js]: We are creating those requests as local variables in a loop but we want their callback function to execute after the loop finishes. I was just worried that they might be garbage collected before their callbacks are called. I figured that putting them inside an array will prevent them from being garbage collected.

Comment (marcinwasowicz) [reply src/drivers/indexeddb.js]: The reasoning was similar for `getMultipleItems`.

**Review by xsanm — APPROVED** [review]

**Pull request merged**